src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,273 @@
+        {
+          /// Would never happen for i == 0
+          upper_index = i - 1;
+          break;
+        }
+      }
+      const size_t random_idx = uniform_distribution_portable(mt, upper_index + 1);
+      const swarm_id_t swarm_id = sorted_swarm_sizes[random_idx].swarm_id;
+      swarm_to_snodes.at(swarm_id).push_back(sn_pk);
+      /// run the excess/threshold round after each additional snode
+      create_new_swarm_from_excess(swarm_to_snodes, mt);
+    }
+  }
+
+  /// This function needs to be called after each single snode is stolen
+  prod_static bool calc_robin_hood_round(const swarm_snode_map_t &swarm_to_snodes, std::vector<excess_pool_snode> &rich_snodes, std::vector<swarm_id_t> &poor_swarm_ids)
+  {
+    std::vector<swarm_size> sorted_swarm_sizes;
+    calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
+    rich_snodes.clear();
+    poor_swarm_ids.clear();
+    const size_t rich_percentile_index = STEALING_SWARM_UPPER_PERCENTILE * (sorted_swarm_sizes.size() - 1) / 100;
+    size_t rich_percentile_value = sorted_swarm_sizes.at(rich_percentile_index).size;
+    /// MIN_SWARM_SIZE + 1 because we don't want to steal from snodes that would starve if we do so
+    rich_percentile_value = std::max(MIN_SWARM_SIZE, rich_percentile_value);
+    size_t deficit = 0;
+    size_t excess = 0;
+    for (const auto &entry : swarm_to_snodes)
+    {
+      const size_t swarm_size = entry.second.size();
+      if (swarm_size < MIN_SWARM_SIZE)
+      {
+        deficit += MIN_SWARM_SIZE - swarm_size;
+        poor_swarm_ids.push_back(entry.first);
+      }
+      else if (swarm_size > rich_percentile_value)
+      {
+        excess += swarm_size - rich_percentile_value;
+        for (const auto &sn_pk : entry.second)
+        {
+          rich_snodes.push_back({sn_pk, entry.first});
+        }
+      }
+    }
+    /// We only try to fill one starving swarm at a time, with 1 snode
+    /// So it's ok if there is more deficit than excess
+    return deficit > 0 && excess > 0;
+  };
+
+  void calc_swarm_changes(swarm_snode_map_t &swarm_to_snodes, uint64_t seed)
+  {
+
+    if (swarm_to_snodes.size() == 0)
+    {
+      // nothing to do
+      return;
+    }
+
+    std::mt19937_64 mersenne_twister(seed);
+
+    std::vector<crypto::public_key> unassigned_snodes;
+    const auto it = swarm_to_snodes.find(UNASSIGNED_SWARM_ID);
+    if (it != swarm_to_snodes.end()) {
+      unassigned_snodes = it->second;
+      swarm_to_snodes.erase(it);
+    }
+
+    LOG_PRINT_L3("calc_swarm_changes. swarms: " << swarm_to_snodes.size() << ", regs: " << unassigned_snodes.size());
+
+    /// 0. Ensure there is always 1 swarm
+    if (swarm_to_snodes.size() == 0)
+    {
+      const auto new_swarm_id = get_new_swarm_id(mersenne_twister, {});
+      swarm_to_snodes.insert({new_swarm_id, {}});
+      LOG_PRINT_L2("Created initial swarm " << new_swarm_id);
+    }
+
+    // TODO?
+    /// Handle snodes in the buffer if any (one time process)
```

Comment:
There will necessarily need to be a full scan so this comment is outdated

Reply:
Will remove.